drivers/accel/adxl367/adxl367.c

```diff
@@ -0,0 +1,1310 @@
+		samples_per_set = 1;
+		break;
+	case ADXL367_FIFO_FORMAT_XYZT:
+	case ADXL367_FIFO_FORMAT_XYZA:
+		samples_per_set = 4;
+		break;
+	case ADXL367_FIFO_FORMAT_XT:
+	case ADXL367_FIFO_FORMAT_YT:
+	case ADXL367_FIFO_FORMAT_ZT:
+	case ADXL367_FIFO_FORMAT_XA:
+	case ADXL367_FIFO_FORMAT_YA:
+	case ADXL367_FIFO_FORMAT_ZA:
+		samples_per_set = 2;
+		break;
+	default:
+		return -1;
+	}
+
+	return 0;
+}
+
+/***************************************************************************//**
+ * @brief Configures the FIFO feature. Uses ADXL367_14B_CHID read mode as
+ * 		default.
+ *
+ * @param dev          - The device structure.
+ * @param mode         - FIFO mode selection.
+ *                       Example: ADXL367_FIFO_DISABLED,
+ *								  ADXL367_OLDEST_SAVED,
+ *								  ADXL367_STREAM_MODE,
+ *								  ADXL367_TRIGGERED_MODE
+ * @param format	   - FIFO format selection.
+ * 				         Example:  ADXL367_FIFO_FORMAT_XYZ,
+ *								   ADXL367_FIFO_FORMAT_X,
+ *								   ADXL367_FIFO_FORMAT_Y,
+ *								   ADXL367_FIFO_FORMAT_Z,
+ *								   ADXL367_FIFO_FORMAT_XYZT,
+ *	 							   ADXL367_FIFO_FORMAT_XT,
+ *								   ADXL367_FIFO_FORMAT_YT,
+ *								   ADXL367_FIFO_FORMAT_ZT,
+ *								   ADXL367_FIFO_FORMAT_XYZA,
+ *								   ADXL367_FIFO_FORMAT_XA,
+ *								   ADXL367_FIFO_FORMAT_YA,
+ *								   ADXL367_FIFO_FORMAT_ZA
+ * @param sets_nb 	   - Specifies the number of samples sets to store in the FIFO.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t adxl367_fifo_setup(struct adxl367_dev *dev,
+			   enum adxl367_fifo_mode mode,
+			   enum adxl367_fifo_format format,
+			   uint8_t sets_nb)
+{
+	int32_t ret;
+
+	ret = adxl367_set_fifo_mode(dev, mode);
+	if (ret)
+		return ret;
+
+	ret = adxl367_set_fifo_format(dev, format);
+	if (ret)
+		return ret;
+
+	ret = adxl367_set_fifo_sample_sets_nb(dev, sets_nb);
+	if (ret)
+		return ret;
+
+	return adxl367_set_fifo_read_mode(dev, ADXL367_14B_CHID);
+}
+
+/***************************************************************************//**
+ * @brief Reads raw values from FIFO. If, after setting FIFO mode, any of x, y, z,
+ * 		temp or adc aren't selected, assign NULL pointer. Uses ADXL367_14B_CHID
+ * 		read mode as default.
+ *
+ * @param dev       - The device structure.
+ * @param x  	    - X axis raw data buffer. If not used, assign NULL.
+ * @param y  	    - Y axis raw data buffer. If not used, assign NULL.
+ * @param z  	    - Z axis raw data buffer. If not used, assign NULL.
+ * @param temp_adc  - Data buffer for temperature or adc raw data. If not
+ * 				used assign NULL.
+ * @param sets_nb   - Number of sample sets to be read.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t adxl367_read_raw_fifo(struct adxl367_dev *dev, int16_t *x, int16_t *y,
```

Comment:
Just an idea here. I would implement this function differently. The hardware FIFO is used to store data after the measurement, without losing any data by making single reads for example. With that in mind, I would expect the user to be more interested in all data in the FIFO rather than a specific number of samples. Using FIFO_ENTRIES you can find out how many data sets are stored in the buffer and return the data together with the fifo entries number such that the user is aware how many data sets are valid from a maximum of 512. 
Using this procedure, you won't have to wait for the data to be available. The user can manage that based on the selected sampling frequency. 

I would not use the delay at all in this case.